src/FunctionApps/DSTP/src/main/kotlin/gov/cdc/prime/phdi/Utilities/HelperFunctions.kt

```diff
@@ -0,0 +1,363 @@
+    BHS is "Batch Header Segment", which defines the start of a batch
+    BTS is "Batch Trailer Segment", which defines the end of a batch
+    
+    The structure of an HL7 Batch looks like this:
+    [FHS] (file header segment) { [BHS] (batch header segment)
+    { [MSH (zero or more HL7 messages)
+    ....
+    ....
+    ....
+    ] }
+    [BTS] (batch trailer segment)
+    }
+    [FTS] (file trailer segment)
+
+    We ignore lines that start with these since we don't want to include them in a message
+    */
+    messageLines.forEach {
+        if (it.startsWith("FHS"))
+            return@forEach
+        if (it.startsWith("BHS"))
+            return@forEach
+        if (it.startsWith("BTS"))
+            return@forEach
+        if (it.startsWith("FTS"))
+            return@forEach
+
+        /*
+            If we reach a line that starts with MSH and we have
+            content in nextMessage, then by definition we have 
+            a full message in nextMessage and need to append it
+            to output. This will not trigger the first time we
+            see a line with MSH since nextMessage will be empty
+            at that time.
+        */
+        if (nextMessage.isNotBlank() && it.startsWith("MSH")) {
+            output.add(nextMessage.toString())
+            nextMessage.clear()
+        }
+
+        // Otherwise, continue to add the line of text to nextMessage
+        if (it.isNotBlank()) {
+            nextMessage.append("$it\r")
+        }
+    }
+    /*
+        Since our loop only adds messages to output when it finds
+        a line that starts with MSH, the last message would never
+        be added. So we explicitly add it here.
+    */
+    if (nextMessage.isNotBlank()) {
+        output.add(nextMessage.toString())
+    }
+    return output
+}
+
+/* 
+   ***********************
+    FHIR SERVER FUNCTIONS
+   ***********************
+*/
+// Connect to the FHIR server and get a bearer token
+public fun getAccessToken(): String? {
+    val tenantId: String = System.getenv("tenant_id")
+    val url: String = "https://login.microsoftonline.com/${tenantId}/oauth2/token"
+
+    val requestBody = StringBuilder("grant_type=client_credentials")
+    // use a HashMap instead of JSONObject for easier iteration through key,value pairs
+    val parameters: MutableMap<String, String> = HashMap()
+    parameters.put("client_id", System.getenv("client_id"))
+    parameters.put("client_secret", System.getenv("client_secret"))
+    parameters.put("resource", System.getenv("fhir_url"))
+    parameters.forEach { (key, value) -> requestBody.append("&${key}=${value}") }
+
+    val client: HttpClient = HttpClient.newHttpClient()
+    val request: HttpRequest = HttpRequest.newBuilder()
+        .uri(URI.create(url))
+        .headers("Content-Type", "application/x-www-form-urlencoded")
+        .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
+        .build()
+
+    val response: HttpResponse<String> = client.send(
```

Comment:
The error handling and logging is currently taking place in the `Convert*ToFHIR` functions, which might not be the best place to do it. The current logging just says "Failed to convert a message to FHIR", which doesn't explain why it failed. I've added in the exception's stack trace to the logging.

Reply:
This does, however, make me realize that it's probably worth implementing better error handling instead of just trying once and logging any failures. We should implement retries and probably store files that errored out somewhere different from those that were deemed invalid.

Reply:
That may be helpful once we figure out what the error rate is - if it's 1-2% then :shrug: - maybe leave it as is for now?